### PR TITLE
Feature/rm 533

### PR DESCRIPTION
**Pull request by yalania**
**Ticket**
https://app.hacknplan.com/p/113092/kanban?categoryId=0&boardId=315785&taskId=533&tabId=basicinfo

**Problem**

We have many files that are extracted from FBX when imported, so everytime whe import this file override destrying changes for materials and polluting the git repository.

**Solution**

Save information for extracted files in library folder, this way everytime we import we are not polluting the assets folder, I added UI for having the files available in the project explorer and also new porperty remapped materials to the model metafile in order to extract materials and map fbx with different materials.

### CODE REVIEW

File: Engine/EditorUI/Panel/InspectorSubpanel/PanelMaterial.cpp [hunk 1]
```diff
@@ -113,13 +122,19 @@ void PanelMaterial::Render(std::shared_ptr<Material> material)
...
 	{
 		ImGui::PopItemFlag();
 		ImGui::PopStyleVar();
 	}
 }
 
+bool PanelMaterial::IsMaterialExtracted(const std::shared_ptr<Material> &material)
+{
+	Metafile * meta = App->resources->resource_DB->GetEntry(material->GetUUID());
+	return meta->version == 0;
```
Comment (yalania) [inline Engine/EditorUI/Panel/InspectorSubpanel/PanelMaterial.cpp:135]: Very tricky this one, but easy way to do it, don't judge me pls xD

Comment (mariofv) [reply Engine/EditorUI/Panel/InspectorSubpanel/PanelMaterial.cpp:135]: òwó

File: Engine/Helper/Config.cpp [hunk 2]
```diff
@@ -1,14 +1,22 @@
...
 
+Config::Config(FileData & data)
+{
+	std::string serialized_string((char*)data.buffer, data.size);
+	free((char*)data.buffer);
+	config_document.Parse(serialized_string.c_str());
+	allocator = &config_document.GetAllocator();
+}
+
 Config::Config(const rapidjson::Value& object_value)
```
Comment (mariofv) [inline Engine/Helper/Config.cpp:20]: Like it :D

File: Engine/ResourceManagement/Importer/ModelImporter.h [hunk 3]
```diff
@@ -28,7 +30,7 @@ struct CurrentModelData
 	Path* asset_file_folder_path = nullptr;
 	float scale = 1.f;
 	std::map<std::string, uint32_t> skeleton_cache;
-	const ModelMetafile* model_metafile;
+ ModelMetafile* model_metafile;
```
Comment (mariofv) [inline Engine/ResourceManagement/Importer/ModelImporter.h:33]: Identation

File: Engine/ResourceManagement/Importer/ModelImporter.h [hunk 4]
```diff
@@ -40,6 +42,7 @@ struct CurrentModelData
 private:
 	std::vector<Config> ExtractDataFromNode(const aiNode* root_node, const aiMatrix4x4& parent_transformation) const;
 	uint32_t ExtractMaterialFromNode(size_t mesh_index, const std::string& mesh_name) const;
+    uint32_t SaveDataInLibrary(Metafile &node_metafile, FileData &mesh_material_data) const;
```
Comment (mariofv) [inline Engine/ResourceManagement/Importer/ModelImporter.h:46]: Identation

File: Engine/EditorUI/Panel/InspectorSubpanel/PanelMaterial.cpp (hunk 1 above)
Comment (Dartemiss) [reply Engine/EditorUI/Panel/InspectorSubpanel/PanelMaterial.cpp:135]: ┗┃・ ■ ・┃┛

File: Engine/ResourceManagement/Importer/PrefabImporter.cpp [hunk 5]
```diff
@@ -12,6 +12,7 @@
 #include "ResourceManagement/Resources/Material.h"
 #include "ResourceManagement/Resources/Mesh.h"
 #include "ResourceManagement/Resources/Skeleton.h"
+#include "ResourceManagement/Metafile/ModelMetafile.h"
```
Comment (Dartemiss) [inline Engine/ResourceManagement/Importer/PrefabImporter.cpp]: Code style xD

**Review by Dartemiss — COMMENTED** [review]
Good job Anabel! I have a question, now I can't modify materials from the inspector, they are like blocked, is this intended? How are materials gonna be modified? Also after this pr is merged we should do another one on the Twinlight repo erasing all .mesh .may files right?

File: Engine/ResourceManagement/Importer/PrefabImporter.cpp (hunk 5 above)
Comment (yalania) [reply Engine/ResourceManagement/Importer/PrefabImporter.cpp]: \>.<

**Comment by yalania** [comment]
For the materials: Actually they should not be modificable at the moment because if you reimport the fbx the materials will be overwritted, for changing materials you have to ways: Extracting the material (right click -> Extract from prefab) or assigning a new material to the metafile in the inspector. I will post a video in discord.

For the assets, Yes we will need to remove the extracted file, the change will work even with these files but they would be trash. I will create a branch there with the same name (feature/RM-533)

**Review by mariofv — APPROVED** [review]
Really good job, I really like the new model importing process and the editors UI 😄

**Review by Dartemiss — APPROVED** [review]

**Pull request merged**